librtas_src/syscall_calls.c

```diff
@@ -1204,6 +1277,39 @@ int rtas_set_sysparm(unsigned int parameter, char *data)
 	return rc ? rc : status;
 }
 
+/**
+ * rtas_set_sysparm
+ * @brief Interface to the ibm,set-system-parameter rtas call
+ *
+ * @param parameter
+ * @param data
+ * @return 0 on success, !0 otherwise
+ */
+int rtas_set_sysparm(unsigned int parameter, /* const */ char *data)
+{
+	int (*set_sysparm_fn)(unsigned int parameter, const char *data);
+	int fd;
+
+	/*
+	 * Use the fallback if:
+	 * - /dev/lparctl can't be opened, or
+	 * - LPARCTL_SET_SYSPARM returns an error for a NULL argument
+	 */
+
+	set_sysparm_fn = set_sysparm_devmem;
+
+	fd = open("/dev/lparctl", O_RDWR | O_CLOEXEC);
+	if (fd != -1) {
+		int res = ioctl(fd, LPARCTL_SET_SYSPARM, NULL);
```

Comment:
Yes, I think you're right that the code as structured in this change could avoid the ioctl with a NULL argument. Just attempt the real command, then fall back if we get `ENOIOCTLCMD` (but not on other errors).

However, for now I'll plan to keep the ability on the kernel side to probe valid commands with a null argument. Eventually this could be used to perform the detection at library init or similar.